wisp/wisp.php

```diff
@@ -597,3 +682,168 @@ function wisp_ClientArea(array $params) {
+        Format:
+        [
+            [<ip_address>] => {
+                [<port number>] => {
+                    ['id'] = 1234;
+                }
+            },
+            ['192.168.1.123'] => {
+                ['1234'] => {
+                    ['id'] = 1234;
+                }z
+            },
+        ]
+    */
+    $available_allocations = array();
+    $allocations = getPaginatedData($params,'nodes/'.$node_id.'/allocations');
+    foreach($allocations as $key => $allocation) {
+        $ip = $allocation['attributes']['ip'];
+        $port = $allocation['attributes']['port'];
+        if ($allocation['attributes']['assigned']!=true) {
+            $available_allocations[$ip][$port]['id'] = $allocation['attributes']['id'];
+        }
+    }
+    return $available_allocations;
+}
+
+function getPaginatedData($params,$url) {
+    /*
+        Makes a paginated API request and returns the response.
+    */
+
+    $results = array();
+    // Fetch and parse first page of data
+    $response = wisp_API($params, $url);
+    foreach($response['data'] as $key => $value) {
+        array_push($results, $value);
+    }
+    $current_page = $response['meta']['pagination']['current_page'];
+    $total_pages = $response['meta']['pagination']['total_pages'];
+    // Fetch and parse any remaining pages
+    while ($total_pages > $current_page) {
+        $next_page = intval($current_page)+1;
+        $response = wisp_API($params, $url.'?page='.$next_page);
+        foreach($response['data'] as $key => $value) {
+            array_push($results, $value);
+        }
+        $current_page = $response['meta']['pagination']['current_page'];
+        $total_pages = $response['meta']['pagination']['total_pages'];
+    }
+    return $results;
+}
+
+function findFreePorts(array $available_allocations, string $port_offsets) {
+    /*
+        This is the main logic that takes a list of available allocations
+        and the required offsets and then finds the first available set.
+        e.g. if port offsets +1 +2 and +4 are requested (format: 1,2,4)
+        we take each port one by one and check if <port number> + 1,
+        <port number> + 2 and <port number> + 4 are available.
+        If all requested port allocations are available, they are returned.
+
+        Inputs:
+        $available_allocations      This is the first port in the range to test.
+                                    All other ports will be checked based on the
+                                    required offset from the first.
+
+        $port_offsets               The json string of offsets from the first port that
+                                    are required for the server.
+
+        Outputs:
+        $ports_found                The array of ports that were found available, based on the offsets
+                                    the additional ports required.
+    */
+    // Iterate over available IP's
+
+    // Decode json string of port offsets
```

Comment:
See the comment in README - this would be redundant with a flipped key-value format.

Reply:
Fixed as per comment in readme 👌 Format has been flipped